crates/aptos-jwk-consensus/src/jwk_observer.rs

```diff
@@ -0,0 +1,136 @@
+// Copyright © Aptos Foundation
+
+use anyhow::Result;
+use aptos_channels::aptos_channel;
+use aptos_logger::{debug, info};
+use aptos_types::jwks::{jwk::JWK, Issuer};
+use futures::{FutureExt, StreamExt};
+use move_core_types::account_address::AccountAddress;
+use serde::{Deserialize, Serialize};
+use std::time::Duration;
+use tokio::{sync::oneshot, task::JoinHandle, time::MissedTickBehavior};
+
+#[derive(Serialize, Deserialize)]
+struct OpenIDConfiguration {
+    issuer: String,
+    jwks_uri: String,
+}
+
+#[derive(Serialize, Deserialize)]
+struct JWKsResponse {
+    keys: Vec<serde_json::Value>,
+}
+
+/// Given an Open ID configuration URL, fetch its JWKs.
+pub async fn fetch_jwks(my_addr: AccountAddress, config_url: Vec<u8>) -> Result<Vec<JWK>> {
+    if cfg!(feature = "smoke-test") {
+        use reqwest::header;
+        let maybe_url = String::from_utf8(config_url);
+        let jwk_url = maybe_url?;
+        let client = reqwest::Client::new();
+        let JWKsResponse { keys } = client
+            .get(jwk_url.as_str())
+            .header(header::COOKIE, my_addr.to_hex())
+            .send()
+            .await?
+            .json()
+            .await?;
+        let jwks = keys.into_iter().map(JWK::from).collect();
+        Ok(jwks)
+    } else {
+        let maybe_url = String::from_utf8(config_url);
+        let config_url = maybe_url?;
+        let client = reqwest::Client::new();
+        let OpenIDConfiguration { jwks_uri, .. } =
+            client.get(config_url.as_str()).send().await?.json().await?;
+        let JWKsResponse { keys } = client.get(jwks_uri.as_str()).send().await?.json().await?;
+        let jwks = keys.into_iter().map(JWK::from).collect();
+        Ok(jwks)
+    }
+}
+
+/// A process thread that periodically fetch JWKs of a provider and push it back to JWKManager.
+pub struct JWKObserver {
+    close_tx: oneshot::Sender<()>,
+    join_handle: JoinHandle<()>,
+}
+
+impl JWKObserver {
+    pub fn spawn(
+        my_addr: AccountAddress,
+        issuer: Issuer,
+        config_url: Vec<u8>,
+        fetch_interval: Duration,
+        observation_tx: aptos_channel::Sender<(), (Issuer, Vec<JWK>)>,
+    ) -> Self {
+        let (close_tx, close_rx) = oneshot::channel();
+        let join_handle = tokio::spawn(Self::thread_main(
+            fetch_interval,
+            my_addr,
+            issuer.clone(),
+            config_url.clone(),
+            observation_tx,
+            close_rx,
+        ));
+        info!(
+            "[JWK] observer spawned, issuer={:?}, config_url={:?}",
+            String::from_utf8(issuer),
+            String::from_utf8(config_url)
+        );
+        Self {
+            close_tx,
+            join_handle,
+        }
+    }
+
+    async fn thread_main(
```

Comment:
you mean main_loop?

Reply:
call `start` to be consistent with the codebase